# Using ImageAI for Image Prediction

This is a mini-project for the Zero-To-Mastery Python Developer course 

Documentation can be found [here](https://github.com/OlafenwaMoses/ImageAI/blob/master/imageai/Classification/README.md)

Using the model MobileNetV2 because it is a lightweight model.

In [1]:
from imageai.Prediction import ImagePrediction
import os

In [2]:
execution_path = os.getcwd()

In [3]:
prediction = ImagePrediction()

## Load MobileNetV2 Model

In [5]:
prediction.setModelTypeAsMobileNetV2()

In [6]:
prediction.setModelPath(os.path.join(execution_path, "mobilenet_v2.h5"))

In [7]:
prediction.loadModel()

In [8]:
predictions, probabilities = prediction.classifyImage(os.path.join(execution_path, "images/giraffe.jpg"), result_count=5 )
for eachPrediction, eachProbability in zip(predictions, probabilities):
    print(eachPrediction , " : " , eachProbability)

40960/35363 [==================================] - 0s 1us/step
leopard  :  1.841278187930584
lynx  :  1.8284115940332413
cheetah  :  1.8218128010630608
jaguar  :  1.6497427597641945
impala  :  0.8935719728469849


In [9]:
# create a function for predictions

def image_prediction(image_name, pred_count=5):
    
    predictions, probabilities = prediction.classifyImage(os.path.join(execution_path, f"images/{image_name}"), result_count=pred_count)
    for eachPrediction, eachProbability in zip(predictions, probabilities):
        print(eachPrediction , " : " , eachProbability)
    

## Make Predictions

In [10]:
image_prediction('godzilla.jpg')

common_iguana  :  21.163669228553772
American_alligator  :  20.356735587120056
pedestal  :  3.057633712887764
triceratops  :  2.950582094490528
African_crocodile  :  1.924213021993637


In [11]:
image_prediction('house.jpg')

boathouse  :  22.32779711484909
church  :  9.003543853759766
palace  :  6.23006671667099
lakeside  :  3.6008939146995544
beacon  :  1.7462363466620445


In [12]:
image_prediction('giraffe.jpg')

leopard  :  1.841278187930584
lynx  :  1.8284115940332413
cheetah  :  1.8218128010630608
jaguar  :  1.6497427597641945
impala  :  0.8935719728469849


## Results:

The predictions are not very good. This is a lightweight model so it is expected that accuracy will be lost due to the smaller model.

### This Giraffe was predicted as Leopard with 1.85 probability

<img src="images/giraffe.jpg" alt="Giraffe"/>

### This Godzilla was predicted as Common Iguana with 21.2 probability

<img src="images/godzilla.jpg" alt="Godzilla"/>

### This House was predicted as Boathouse with 22.33 probability

<img src="images/house.jpg" alt="House"/>